### ISSUE TITLE
Strict key checking issues in v1.16

### ISSUE THREAD

**Issue by menma1234** [taverntesting/tavern#707]
It appears that https://github.com/taverntesting/tavern/pull/604 had a couple of unintended side effects:

1. The default strictness setting appears to have been changed to `off` instead of the expected `on` setting
2. Having the strict level at `list_any_order` no longer does strict key checking within nested objects

As a quick example, I have a dummy server that returns the following JSON:

```json
{
  "a": 1,
  "b": {
    "first": 10,
    "second": 20
  },
  "c": 2
}
```

Tavern file:

```yaml
---
test_name: strict test
stages:
  - name: Simple get
    request:
      url: http://localhost:8080
      method: GET
    response:
      status_code: 200
      json:
        a: 1
        b:
          first: 10
#          second: 20
        c: 2
```

This unexpectedly passes:

```
$ pytest -- test.tavern.yaml
============================= test session starts ==============================
platform linux -- Python 3.9.6, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /volume/temp
plugins: tavern-1.16.0
collected 1 item

test.tavern.yaml .                                                       [100%]

============================== 1 passed in 0.18s ===============================
```

Having `strict: -json:list_any_order` has the same result.

Adding `strict: - json:on` to the YAML causes it to fail, as I would've expected without having to specify it explicitly:

```
$ pytest -- test.tavern.yaml
============================= test session starts ==============================
platform linux -- Python 3.9.6, pytest-6.2.5, py-1.10.0, pluggy-1.0.0
rootdir: /volume/temp
plugins: tavern-1.16.0
collected 1 item

test.tavern.yaml F                                                       [100%]

=================================== FAILURES ===================================
__________________ /volume/temp/test.tavern.yaml::strict test __________________
Format variables:

Source test stage (line 5):
  - name: Simple get
    request:
      url: http://localhost:8080
      method: GET
    response:
      status_code: 200
      json:
        a: 1
        b:
          first: 10

        c: 2

Formatted stage:
  name: Simple get
  request:
    method: GET
    url: http://localhost:8080
  response:
    json:
      a: 1
      b:
        first: 10
      c: 2
    status_code: 200

Errors:
E   tavern.util.exceptions.TestFailError: Test 'Simple get' failed:
    - Structure of returned data was different than expected  - Extra keys in response: {'second'} (expected["b"] = '{'first': 10}' (type = <class 'dict'>), actual["b"] = '{'first': 10, 'second': 20}' (type = <class 'dict'>))
------------------------------ Captured log call -------------------------------
ERROR    tavern.response.base:base.py:41 Structure of returned data was different than expected  - Extra keys in response: {'second'} (expected["b"] = '{'first': 10}' (type = <class 'dict'>), actual["b"] = '{'first': 10, 'second': 20}' (type = <class 'dict'>))
Traceback (most recent call last):
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/util/dict_util.py", line 402, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert {'a': 1, 'b':...: 20}, 'c': 2} == {'a': 1, 'b':...: 10}, 'c': 2}
  Omitting 2 identical items, use -vv to show
  Differing items:
  {'b': {'first': 10, 'second': 20}} != {'b': {'first': 10}}
  Use -v to get the full diff

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/util/dict_util.py", line 402, in check_keys_match_recursive
    assert actual_val == expected_val
AssertionError: assert {'first': 10, 'second': 20} == {'first': 10}
  Omitting 1 identical items, use -vv to show
  Left contains 1 more item:
  {'second': 20}
  Use -v to get the full diff

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/response/base.py", line 89, in recurse_check_key_match
    check_keys_match_recursive(expected_block, block, [], strict)
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/util/dict_util.py", line 460, in check_keys_match_recursive
    check_keys_match_recursive(
  File "/volume/temp/.venv/lib/python3.9/site-packages/tavern/util/dict_util.py", line 445, in check_keys_match_recursive
    raise exceptions.KeyMismatchError(full_msg) from e
tavern.util.exceptions.KeyMismatchError: Structure of returned data was different than expected  - Extra keys in response: {'second'} (expected["b"] = '{'first': 10}' (type = <class 'dict'>), actual["b"] = '{'first': 10, 'second': 20}' (type = <class 'dict'>))
=========================== short test summary info ============================
FAILED test.tavern.yaml::strict test
============================== 1 failed in 0.21s ===============================
```

If I change the Tavern YAML to comment out the `c` key from the response:

```yaml
---
test_name: strict test
stages:
  - name: Simple get
    request:
      url: http://localhost:8080
      method: GET
    response:
      status_code: 200
      json:
        a: 1
        b:
          first: 10
#          second: 20
#        c: 2
```

This passes without any `strict` key and fails in both `list_any_order` and `on`.

**Comment by michaelboulton** [comment]
This actually appears to be a bug in `extract_strict_setting` and the associated parsing in dict_util which doesn't take into account the fact that strictness being 'true' or 'false' needs to take what is being checked into account, so the strict checking works on the top level keys but not on any nested keys. Hopefully this shouldn't be too difficult to fix

**Comment by michaelboulton** [comment]
Should be fixed in 1.16.1
